interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+type componentType int
+
+// Component types in order from lowest to highest precedence.
+const (
+	compUnset componentType = iota
+	compGlobstar
+	compSeparatorDoublestarTerminal          // need to bundle separator and terminal marker with /**
+	compSeparatorDoublestarSeparatorTerminal // need to bundle separators and terminal marker with /**/
+	compSeparatorDoublestar
+	compTerminal // marker of end of pattern.
+	compSeparator
+	compAnySingle // ? has precedence over / so that /foo*?/ has precedence over /foo*/
+	compLiteral
+)
+
+type component struct {
+	compType componentType
+	compText string
+}
+
+func (c *component) String() string {
+	switch c.compType {
+	case compGlobstar:
+		return "*"
+	case compSeparatorDoublestarTerminal:
+		return "/**"
+	case compSeparatorDoublestarSeparatorTerminal:
+		return "/**/"
+	case compSeparatorDoublestar:
+		return "/**"
+	case compTerminal: // end of pattern
+		return ""
+	case compSeparator:
+		return "/"
+	case compAnySingle:
+		return "?"
+	case compLiteral:
+		return c.compText
+	}
+	return "###ERR_UNKNOWN_COMPONENT_TYPE###" // Should not occur
+}
+
+func (c *component) componentRegex() string {
+	switch c.compType {
+	case compGlobstar:
+		return `((?:[^/]|\\/)*)`
+	case compSeparatorDoublestarTerminal:
+		return `((?:/.+)?/?)`
+	case compSeparatorDoublestarSeparatorTerminal:
+		return `((?:/.+)?/)`
+	case compSeparatorDoublestar:
+		return `((?:/.+)?)`
+	case compTerminal:
+		return `(/?)`
```

Comment:
Should this have $ somewhere?

Reply:
The composite regexp is always surrounded by `^` and `$`:

https://github.com/snapcore/snapd/pull/13868/files#diff-93d4ef509daf0fd8b355ce49f9a963e8ae2d819854995c7fa86d30956ae438beR248-R256